### ISSUE TITLE
Error parsing relay mutation

### ISSUE THREAD

**Issue by dre1080** [absinthe-graphql/absinthe#62]
When I try to parse a Relay mutation, I get the following exception:

```
** (FunctionClauseError) no function clause matching in :absinthe_parser.extract_binary/1
    (absinthe) src/absinthe_parser.yrl:210: :absinthe_parser.extract_binary("input")
    (absinthe) src/absinthe_parser.yrl:105: :absinthe_parser.yeccpars2_72/7
    (absinthe) /opt/boxen/homebrew/Cellar/erlang/18.2.1/lib/erlang/lib/parsetools-2.1.1/include/yeccpre.hrl:57: :absinthe_parse
r.yeccpars0/5
    (absinthe) lib/absinthe.ex:177: Absinthe.parse/1
```

Heres my schema:

```
object :user do
    field :id, :id
    field :email, :string
    field :name, :string
  end

input_object :create_user_input do
    field :email, non_null(:string)
    field :name, non_null(:string)
end

mutation do
    @desc "Create a user"
    field :create_user, type: :user do
      arg :input, non_null(:create_user_input)

      resolve &Resolver.User.create/2
    end
end
```

Relay submits a mutation like:
```
mutation CreateUserMutation($input_0:createUserInput!){createUser(input:$input_0){clientMutationId}}
```

**Comment by bruce** [comment]
Thanks, @dre1080 -- this is very helpful. We'll get on this.

**Comment by bruce** [comment]
It looks like this is due to the parser (or, perhaps more accurately, the lexer) choking on the `input` argument name, as `input` is a reserved word. This definitely seems like a bug -- argument names are unambiguously argument names, and there's nothing in the GraphQL specification that would disallow, eg, `input` as an argument name. I'm working on fixing it.

**Comment by dre1080** [comment]
Ahh I see. There's no way I can change the `input` argument name as Relay requires it for mutations.

**Comment by dre1080** [comment]
@bruce After testing on master this is the exception im now getting:

```
** (FunctionClauseError) no function clause matching in Absinthe.Type.valid_input?/2
        (absinthe) lib/absinthe/type.ex:192: Absinthe.Type.valid_input?(%Absinthe.Type.InputObject{__reference__: %{identifier:
 :create_user_input, location: %{file: "/web/graph/schema/type/user.ex", line: 22}, module: Gr
aph.Schema.Type.User}, description: nil, fields: %{email: %Absinthe.Type.Field{__reference__: %{identifier: :email, location: %
{file: "/web/graph/schema/type/user.ex", line: 23}, module: Graph.Schema.Type.User}, args: %{}
, default_value: nil, deprecation: nil, description: nil, name: "email", resolve: nil, type: %Absinthe.Type.NonNull{of_type: :s
tring}}, name: %Absinthe.Type.Field{__reference__: %{identifier: :name, location: %{file: "/web/graph/s
chema/type/user.ex", line: 24}, module: Graph.Schema.Type.User}, args: %{}, default_value: nil, deprecation: nil, desc
ription: nil, name: "name", resolve: nil, type: %Absinthe.Type.NonNull{of_type: :string}}}, name: "CreateUserInput"}, %{"client
MutationId" => "0"})
        (absinthe) lib/absinthe/execution/variables.ex:49: Absinthe.Execution.Variables.do_parse/5
        (elixir) lib/enum.ex:1473: Enum."-reduce/3-lists^foldl/2-0-"/3
        (absinthe) lib/absinthe/execution.ex:176: Absinthe.Execution.set_variables/1
        (absinthe) lib/absinthe/execution.ex:29: Absinthe.Execution.run/2
        (absinthe) lib/absinthe.ex:238: Absinthe.run/3
        (absinthe_plug) lib/absinthe/plug.ex:78: Absinthe.Plug.do_call/5
```

**Comment by bruce** [comment]
Reopening to work through any other issues with this request.

@dre1080 Let me know how thing go against master now.

**Comment by dre1080** [comment]
Thanks..

This is what I'm getting now:

```
** (MatchError) no match of right hand side value: %Absinthe.Type.InputObject{__reference__: %{identifier: :create_user_inp
ut, location: %{file: "/web/graph/schema/type/user.ex", line: 22}, module: Graph.Schema.Type.U
ser}, description: nil, fields: %{email: %Absinthe.Type.Field{__reference__: %{identifier: :email, location: %{file: "/web/graph/schema/type/user.ex", line: 23}, module: Graph.Schema.Type.User}, args: %{}, default_value:
nil, deprecation: nil, description: nil, name: "email", resolve: nil, type: %Absinthe.Type.NonNull{of_type: :string}}, name: %A
bsinthe.Type.Field{__reference__: %{identifier: :name, location: %{file: "/web/graph/schema/type/user.e
x", line: 24}, module: Graph.Schema.Type.User}, args: %{}, default_value: nil, deprecation: nil, description: nil, nam
e: "name", resolve: nil, type: %Absinthe.Type.NonNull{of_type: :string}}}, name: "CreateUserInput"}
        (absinthe) lib/absinthe/execution/variables.ex:95: Absinthe.Execution.Variables.coerce/2
        (absinthe) lib/absinthe/execution/variables.ex:66: Absinthe.Execution.Variables.valid/4
```

**Comment by bruce** [comment]
@dre1080 Can you share a bit more of what your schema looks like at the moment? This is the same request as before, to verify?

**Comment by dre1080** [comment]
@bruce Yes, same request

**Comment by bruce** [comment]
We use InputObjects all the time in our own schema, so wondering what's going on here. Hmmm...

**Comment by benwilson512** [comment]
Can you show us some information about what your request URI and body look like?

**Comment by dre1080** [comment]
This is my request payload
```
{
query: "mutation createUserMutation($input_0:createUserInput){createUser(input:$input_0){clientMutationId}}"
variables: {input_0: {clientMutationId: "0", email: "john@example.com", name: "John"}}
}
```

**Comment by benwilson512** [comment]
Hey! So we've spent some more time looking at this and it seems that you've actually run into a separate problem unrelated to this original issue.

We're gonna close this issue and create a new one related to the new problem, which has to do with using input objects as variables.

Thanks!

**Comment by bruce** [comment]
Closed in favor of #65.
